### PR TITLE
Workaround Intel 16 exception specification of virtual destructor

### PR DESCRIPTION
**Pull request by dalg24**
(Re)Tentative fix of #2772 

NOTE: I assumed only `Kokkos_HostSpace.hpp` and `Kokkos_HBWSpace.hpp` needed the workaround and ignored `Kokkos_CudaSpace.hpp`, `Kokkos_HIP_Space.hpp`, and `Kokkos_OpenMPTargetSpace.hpp`.

### CODE REVIEW

**Comment by masterleinad** [comment]
Let's also match the exception specifier for the definitions to avoid warnings such as in https://godbolt.org/z/BFgTjX.

Also, I don't see a problem with using `noexcept` unconditionally here.

**Comment by dalg24** [comment]
> Let's also match the exception specifier for the definitions to avoid warnings such as in https://godbolt.org/z/BFgTjX.

No warning with Intel.  But sure I can add it to the definition too.

> Also, I don't see a problem with using `noexcept` unconditionally here.

Do you have any concern with the changes I suggest?

**Comment by dalg24** [comment]
It is a compiler bug.  The standard says the [specifier is not needed](http://eel.is/c++draft/except.spec#3).

I can add the specifier if you feel strongly about it.  Or let's see what others think.

**Comment by masterleinad** [comment]
No, I don't feel strongly about it. It just seems to be less clutter (even if we might remove it at some point again).

**Comment by ndellingwood** [comment]
@dalg24 @masterleinad This is showing up in intel/17.0.098 nightly builds as well (they had not completed when I posted this issue).

**Comment by dalg24** [comment]
> This is showing up in intel/17.0.098 nightly builds as well (they had not completed when I posted this issue).

Exact same error message?!  This was not showing when I tried on godbolt with `icc 17.0.0`

**Comment by ndellingwood** [comment]
> Exact same error message?!

Yes, I can investigate a bit more since this isn't showing up for you on godbolt

File: core/src/Kokkos_Macros.hpp [hunk 1]
```diff
@@ -356,8 +356,8 @@
 #endif
 #endif
 
-#if (1700 > KOKKOS_COMPILER_INTEL)
```
Comment (masterleinad) [inline core/src/Kokkos_Macros.hpp]: I guess we could just enable it for `Intel` in general to be on the safe side?

Comment (dalg24) [reply core/src/Kokkos_Macros.hpp]: I was not able to reproduce the error with Intel 17.  My answer is yes we could but I am not changing unless someone request it.

File: core/src/Kokkos_Macros.hpp [hunk 2]
```diff
@@ -356,6 +356,10 @@
 #endif
 #endif
 
+#if (1800 > KOKKOS_COMPILER_INTEL)
```
Comment (masterleinad) [inline core/src/Kokkos_Macros.hpp]: This can be
```suggestion
#if (1800 >= KOKKOS_COMPILER_INTEL)
```

**Comment by masterleinad** [comment]
> Exact same error message?! This was not showing when I tried on godbolt with `icc 17.0.0`

I can reproduce on godbolt if we don't pass `-std=c++11` (https://godbolt.org/z/hvgkea). At least everything seems to look good for ICC 18 and above.

File: core/src/Kokkos_Macros.hpp (hunk 2 above)
Comment (masterleinad) [inline core/src/Kokkos_Macros.hpp]: This is OK but I find it harder to read this way than swapping the two operands (and making sure that `KOKKOS_COMPILER_INTEL` is defined).

Comment (dalg24) [reply core/src/Kokkos_Macros.hpp]: https://github.com/kokkos/kokkos/blob/5f8f405fa4f37422fd41972812b77973f3554d97/core/src/Kokkos_Macros.hpp#L307-L366

**Comment by ndellingwood** [comment]
I'll test this now in the failing configurations

**Comment by dalg24** [comment]
> I'll test this now in the failing configurations

Thanks

**Comment by ndellingwood** [comment]
This seems to have resolved the `SharedAllocationRecord` but then hit the `sentinel` related error addressed by PR #2776; let me merge that change in as well and I'll post back.

**Comment by ndellingwood** [comment]
@dalg24 I confirm this PR, coupled with #2776, fixes the compilation errors in the failing intel/15.0.2 intel/16.0.3 and intel/17.0.098 nightly configurations (this PR alone addressed the `SharedAllocationRecord` error, #2776 was necessary for `sentinel` related errors that began occurring following fixes in this PR).

**Comment by ndellingwood** [comment]
@dalg24 as regards nightly tests this PR is ready for merge.

**Pull request merged**